frontend/src/components/Contexts/MissionListsContext.tsx

```diff
@@ -31,9 +39,77 @@ const fetchMissions = (params: {
     const [ongoingMissions, setOngoingMissions] = useState<Mission[]>([])
     const [missionQueue, setMissionQueue] = useState<Mission[]>([])
+    const { registerEvent, connectionReady } = useSignalRContext()
+
+    useEffect(() => {
+        if (connectionReady) {
+            registerEvent('mission run created', (username: string, message: string) => {
+                const newMission: Mission = JSON.parse(message)
+                if (missionQueue.find((m) => m.id === newMission.id))
+                    setMissionQueue((oldQueue) => [...oldQueue, newMission])
+                else
+                    setMissionQueue((oldQueue) => {
+                        let missionQueueCopy = [...oldQueue]
+                        missionQueueCopy = upsertList(missionQueueCopy, newMission)
+                        return [...missionQueueCopy]
+                    })
+            })
+            registerEvent('mission run updated', (username: string, message: string) => {
+                let updatedMission: Mission = JSON.parse(message)
+                // This conversion translates from the enum as a number to an enum as a string
+                updatedMission.status = Object.values(MissionStatus)[updatedMission.status as unknown as number]
+
+                setMissionQueue((oldQueue) => {
+                    const oldQueueCopy = [...oldQueue]
+                    const existingMissionIndex = oldQueue.findIndex((m) => m.id === updatedMission.id)
+                    if (existingMissionIndex !== -1) {
+                        if (updatedMission.status !== MissionStatus.Pending)
+                            oldQueueCopy.splice(existingMissionIndex, 1)
+                        else oldQueueCopy[existingMissionIndex] = updatedMission
+                    }
+                    return oldQueueCopy
+                })
+                setOngoingMissions((oldQueue) => {
+                    const oldQueueCopy = [...oldQueue]
+                    const existingMissionIndex = oldQueue.findIndex((m) => m.id === updatedMission.id)
+                    if (
+                        updatedMission.status === MissionStatus.Ongoing ||
+                        updatedMission.status === MissionStatus.Paused
+                    ) {
+                        if (existingMissionIndex !== -1) {
+                            // Mission is ongoing and in the queue
+                            oldQueueCopy[existingMissionIndex] = updatedMission
+                        } else {
+                            // Mission is ongoing and not in the queue
+                            return [...oldQueueCopy, updatedMission]
+                        }
+                    } else {
+                        if (existingMissionIndex !== -1) {
+                            // Mission is not ongoing and in the queue
+                            oldQueueCopy.splice(existingMissionIndex, 1)
+                        }
+                    }
```

Comment:
Can this logic be hidden in a well named function?

Reply:
I've split them into new functions, hopefully you find them well named